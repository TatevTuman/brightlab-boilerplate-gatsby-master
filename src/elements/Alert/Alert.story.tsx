import React from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'
import { Provider as AlertProvider, transitions, useAlert, AlertProviderProps } from 'react-alert'
import Alert, { AlertProps } from './Alert'
import { Button, ButtonSizes } from '@elements'

interface AlertStoryTemplateProps extends Omit<AlertProviderProps, 'template'> {
  type: AlertProps['options']['type']
  onClose: AlertProps['options']['onClose']
  onOpen: AlertProps['options']['onOpen']
  message: AlertProps['message']
  close: AlertProps['close']
  style: AlertProps['style']
}

const ALERT_PROVIDER_POSITIONS = [
  'top left',
  'top center',
  'top right',
  'middle left',
  'middle',
  'middle right',
  'bottom left',
  'bottom center',
  'bottom right'
] as const

export default {
  title: 'Elements/Alert',
  argTypes: {
    message: { control: 'text', name: 'Message' },
    position: {
      control: {
        type: 'select',
        options: ALERT_PROVIDER_POSITIONS
      },
      name: 'Position'
    },
    type: { control: { type: 'select', options: ['success', 'info', 'error'] }, name: 'Type' },
    transition: { control: { type: 'select', options: ['fade', 'scale'] }, name: 'Transition' },
    timeout: { control: 'number', name: 'Timeout' },
    offset: { control: 'text', name: 'Offset' },
    style: { control: 'object', name: 'Style' },
    onOpen: { action: 'onOpen' },
    onClose: { action: 'onClose' }
  }
} as Meta

const AlertButton: React.FC<AlertProps> = props => {
  const { message, options } = props
  const alert = useAlert()

  const handleAlertShow = () => alert.show(message, options)

  return (
    <Button onClick={handleAlertShow}>
      Show
    </Button>
  )
}

const AlertStoryTemplate: Story<AlertStoryTemplateProps> = args => {
  const { id, style, message, close, type, onOpen, onClose, ...alertProviderProps } = args

  const alertOptions = {
    style,
    message,
    close,
    options: {
      type,
      onOpen,
      onClose,
      position: alertProviderProps.position
    }
  }

  return (
    <AlertProvider template={Alert} {...alertProviderProps}>
      <AlertButton {...alertOptions} />
    </AlertProvider>
  )
}

AlertStoryTemplate.args = {
  position: ALERT_PROVIDER_POSITIONS[4],
  timeout: 2000,
  offset: '0px 0px 0px 0px',
  transition: transitions.SCALE,
  style: {},
  type: 'success',
  message: 'Text',
  onOpen: action('onOpen'),
  onClose: action('onClose')
}

export const AlertStory = AlertStoryTemplate.bind({})
AlertStory.args = {
  ...AlertStoryTemplate.args
}
