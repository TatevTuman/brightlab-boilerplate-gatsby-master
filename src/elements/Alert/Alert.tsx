import React from 'react'
import { AlertComponentPropsWithStyle } from 'react-alert'

export interface AlertProps extends Omit<AlertComponentPropsWithStyle, 'id'> {}

// the style contains only the margin given as offset
// options contains all alert given options
// message is the alert message
// close is a function that closes the alert
const Alert: React.FC<AlertProps> = props => {
  const { style, options = {}, message, close } = props
  const { type } = options

  return (
    <div className={'alert'} data-type={type} style={style} onClick={close} data-testid={'alert'}>
      {message}
    </div>
  )
}

Alert.defaultProps = {
  style: {},
  message: 'Alert message',
  close: () => null,
  options: {
    onClose: () => null,
    onOpen: () => null,
    timeout: 2000
  }
}

export default Alert
