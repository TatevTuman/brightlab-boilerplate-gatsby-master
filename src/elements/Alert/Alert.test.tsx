import React from 'react'
import { AlertType } from 'react-alert'
import { render, cleanup, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import Alert from './Alert'

jest.useFakeTimers()

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('Alert', () => {
  const props = {
    id: 'test',
    message: 'Test message',
    options: {
      onClose: jest.fn(),
      onOpen: jest.fn(),
      timeout: 1000
    },
    close: jest.fn(),
    style: {}
  }

  it('renders correctly', async () => {
    const { container, getByText } = render(<Alert {...props} />)
    await waitFor(() => container)

    expect(container).toMatchSnapshot()
    expect(getByText(props.message)).toBeInTheDocument()
  })

  it('props[type] works correctly', async () => {
    const { getByTestId, rerender } = render(<Alert {...props} />)
    const types: AlertType[] = ['success', 'info', 'error']

    types.forEach(type => {
      rerender(<Alert {...props} options={{ type }} />)
      const alert = getByTestId('alert')
      expect(alert).toHaveAttribute('data-type', type)
    })
  })

  it('props[close] works correctly', async () => {
    const { getByTestId } = render(<Alert {...props} />)
    const alert = getByTestId('alert')

    userEvent.click(alert!)

    expect(props.close).toHaveBeenCalledTimes(1)
  })
})
