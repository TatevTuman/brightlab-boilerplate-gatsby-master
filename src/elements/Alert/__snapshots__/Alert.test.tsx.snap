// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert renders correctly 1`] = `
<div>
  <div
    class="alert"
    data-testid="alert"
  >
    Test message
  </div>
</div>
`;
